orttraining/orttraining/test/graph/module_gradient_graph_builder_test.cc

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <fstream>
+
+#include "gtest/gtest.h"
+#include "test/framework/test_utils.h"
+#include "orttraining/core/framework/module_gradient_graph_builder.h"
+
+namespace onnxruntime {
+namespace test {
+
+void RunModuleGradientGraphBuilderTest(const std::string& file_path,
+                                       const std::vector<std::string>& initializer_names_to_train,
+                                       const std::map<std::string, int>& expected_forward_ops_count,
+                                       const std::map<std::string, int>& expected_backward_ops_count) {
+  onnxruntime::training::ModuleGradientGraphBuilderConfiguration config;
+  config.initializer_names_to_train.assign(initializer_names_to_train.begin(), initializer_names_to_train.end());
+  std::ifstream model_istream(file_path, std::ifstream::in | std::ifstream::binary);
+  onnxruntime::training::ModuleGradientGraphBuilder module_gradient_graph_builder;
+  ASSERT_TRUE(module_gradient_graph_builder.Initialize(model_istream, config).IsOK());
+  ASSERT_TRUE(module_gradient_graph_builder.BuildAndSplit().IsOK());
+
+  // Forward graph.
+  std::istringstream forward_is(module_gradient_graph_builder.GetForwardModel());
+  ONNX_NAMESPACE::ModelProto forward_model_proto;
+  ASSERT_TRUE(Model::Load(forward_is, &forward_model_proto).IsOK());
+  Model forward_model(forward_model_proto, nullptr, logging::LoggingManager::DefaultLogger());
+  std::map<std::string, int> actual_forward_ops_count = CountOpsInGraph(forward_model.MainGraph());
+  ASSERT_EQ(actual_forward_ops_count.size(), expected_forward_ops_count.size());
+  for (const auto& op_count : actual_forward_ops_count) {
+    ASSERT_TRUE(expected_forward_ops_count.find(op_count.first) != expected_forward_ops_count.end());
+    ASSERT_EQ(op_count.second, expected_forward_ops_count.at(op_count.first));
+  }
+
+  std::cout << std::endl;
+
+  // Backward graph.
+  std::istringstream backward_is(module_gradient_graph_builder.GetBackwardModel());
+  ONNX_NAMESPACE::ModelProto backward_model_proto;
+  ASSERT_TRUE(Model::Load(backward_is, &backward_model_proto).IsOK());
+  Model backward_model(backward_model_proto, nullptr, logging::LoggingManager::DefaultLogger());
+  std::map<std::string, int> actual_backward_ops_count = CountOpsInGraph(backward_model.MainGraph());
+  ASSERT_EQ(actual_backward_ops_count.size(), expected_backward_ops_count.size());
+  for (const auto& op_count : actual_backward_ops_count) {
+    ASSERT_TRUE(expected_backward_ops_count.find(op_count.first) != expected_backward_ops_count.end());
+    ASSERT_EQ(op_count.second, expected_backward_ops_count.at(op_count.first));
+  }
+
+  // SplitGraphsInfo.
```

Comment:
it's great that we are matching actual graph inputs/outputs with split_graphs_info, but split_graphs_info is also generated in the module_gradient_graph_builder. 

maybe we should also have some check on the correctness of split_graphs_info itself.